### PR TITLE
TP support for reverse KL loss 

### PR DESCRIPTION
**Pull request by oleksost**
TP support for reverse KL loss.
- adds support for vocabulary parallel reverse KL loss calculation using torch (no fused implementataion). 
- Sequence parallel loss calculation is not supported to keep it simple (I don't think we use sequence parallel embeddings/head)
- this also fixes a small bug in CE loss for when it is used for distillation

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

- added _torch_reverse_kl_forward_backward in cross_entropy.py
- added test_rkl_loss

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [x] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/contributing/contributing).
- [x] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [x] 🎉 The functionality is complete, and I have tested the changes.
- [ ] 📝 I have updated the documentation if needed.
- [ ] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [ ] 🧩 I have commented my code, especially in hard-to-understand areas.

### Dependencies and Configuration

- [ ] 🐋 I have updated the Docker configuration or dependencies, if applicable.
- [ ] 🔄 I have ensured compatibility with the existing setup after dependency changes.

### Testing

- [x] 🧪 I have added or updated tests to cover my changes.
- [x] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

## 📊 Performance Impact Details

If there is any impact on performance, describe it and provide benchmark results, if applicable:

---

## 🗒️ Additional Notes

Include any additional context, information, or considerations here, such as known issues, follow-up tasks, or backward compatibility concerns.

### CODE REVIEW

File: fast_llm/functional/cross_entropy.py [hunk 1]
```diff
@@ -145,13 +145,19 @@ def _fused_cross_entropy_forward_backward(
     else:
         predicted_logits = (target * logits_norm).sum(dim=-1, keepdim=True)
 
+    # shouldn't the predicted_logits be scaled by the number of ranks so that the average loss is correct? i.e.
```
Comment (jlamypoirier) [inline fast_llm/functional/cross_entropy.py]: Not sure what you mean here. We sum over the contributions from each TP rank, so the result is the total predicted logits, and scaling it doesn't make sense.

File: fast_llm/functional/config.py [hunk 2]
```diff
@@ -89,6 +89,12 @@ def _set_activation_fn_map() -> None:
 MAX_DROPLESS_BLOCK_SIZE_ROW = 128
 
 
+class ReverseKLImpl(str, enum.Enum):
```
Comment (jlamypoirier) [inline fast_llm/functional/config.py]: Not needed, there is only one implementation.

File: tests/test_distillation_loss.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+
+@pytest.mark.parametrize("use_mask", [True, False])
+def test_reverse_kl_vocab_tp_two_ranks(use_mask):
+    _spawn_dist(2, _vocab_tp_worker, use_mask)
+
+
+@pytest.mark.parametrize("use_mask", [True, False])
+def test_cross_entropy_vocab_tp_two_ranks(use_mask):
+    _spawn_dist(2, _ce_vocab_tp_worker, use_mask)
```
Comment (jlamypoirier) [inline tests/test_distillation_loss.py]: Need to mark as slow (how long does this take?). Does it work on cpu? Also it would be better to run all tests in the same spawn call because of the huge distributed overhead.

File: tests/test_distillation_loss.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            target_format=TargetFormat.logits,
+            sequence_parallel_logits=False,
+        )
+    else:
+        ref_loss = torch.zeros_like(loss)
+    dist.broadcast(ref_loss, src=0, group=group)
+    torch.testing.assert_close(loss, ref_loss, atol=1e-6, rtol=1e-6)
+
+
+def _ce_vocab_tp_worker(rank: int, group: dist.ProcessGroup, use_mask: bool):
```
Comment (jlamypoirier) [inline tests/test_distillation_loss.py]: We might want to match the implementation and parametrization from `test_cross_entropy`

File: fast_llm/functional/cross_entropy.py [hunk 5]
```diff
@@ -144,14 +144,20 @@ def _fused_cross_entropy_forward_backward(
             all_reduce(predicted_logits, op=ReduceOp.SUM, group=group)
     else:
         predicted_logits = (target * logits_norm).sum(dim=-1, keepdim=True)
+    if group is not None:
+        # this is needed because on each rank we calculate log Z - sum_i t_i * z_i, z_i is logit.
+        # then we average: 1/K sum_ranks (log Z - sum_i t_i * z_i)
+        # = log Z - 1/K sum_ranks (sum_i t_i * z_i)
+        # but sum_ranks (sum_i t_i * z_i) = sum_i t_i * z_i (over all vocab)
+        predicted_logits = predicted_logits * group.size()
```
[the remark this answers is not in the document]
Comment (oleksost) [reply fast_llm/functional/cross_entropy.py]: Yeh, either of two
- scale `predicted_logits` by group size and keep everything as is (i.e. still AVG reduction on loss)
- or do SUM reduction on `predicted_logits` instead of AVG reduction on loss below

**Review by jlamypoirier — APPROVED** [review]
Looks good, but some suggestions on improving the tests
